File: cases/core/synthetic_vulnerable/RS-SV-001/project/src/agent/router.rs
//! Agent router — dispatches plan steps to the appropriate tool.

use std::path::Path;

use crate::agent::planner::PlanStep;
use crate::config::{Config, ALLOWED_EXTENSIONS};
use crate::tools::file_writer::{write_file, WriteResult};

/// Check whether the relative path has an extension in the allowlist.
fn has_allowed_extension(relative_path: &str) -> bool {
    let path = Path::new(relative_path);
    match path.extension().and_then(|ext| ext.to_str()) {
        Some(ext) => {
            let dotted = format!(".{ext}");
            ALLOWED_EXTENSIONS.contains(&dotted.as_str())
        }
        None => false,
    }
}

/// Execute every step in the refactoring plan and collect results.
pub fn execute_plan(plan: &[PlanStep], config: &Config) -> Vec<WriteResult> {
    let mut results = Vec::new();

    for step in plan {
        match step.action.as_str() {
            "rewrite_file" => {
                println!(
                    "[router] rewrite_file: {} — {}",
                    step.relative_path, step.description
                );

                // Enforce the configured file-size limit before writing.
                if step.content.len() > config.max_file_size {
                    results.push(WriteResult {
                        path: step.relative_path.clone(),
                        status: "error".into(),
                        message: format!(
                            "Content size ({} bytes) exceeds maximum ({} bytes)",
                            step.content.len(),
                            config.max_file_size,
                        ),
                    });
                    continue;
                }

                // Only allow writing to files with an approved extension.
                if !has_allowed_extension(&step.relative_path) {
                    results.push(WriteResult {
                        path: step.relative_path.clone(),
                        status: "error".into(),
                        message: "File extension not in allowlist".into(),
                    });
                    continue;
                }

                let result = write_file(
                    &config.workspace_root,
                    &step.relative_path,
                    &step.content,
                );
                results.push(result);
            }
            other => {
                eprintln!("[router] Unknown action: {other}, skipping");
                results.push(WriteResult {
                    path: step.relative_path.clone(),
                    status: "skipped".into(),
                    message: format!("Unknown action: {other}"),
                });
            }
        }
    }

    results
}

File: cases/core/synthetic_vulnerable/PY-SV-003/pr/base/agent/router.py
"""Agent router — dispatches tasks to the appropriate tool."""

from __future__ import annotations

import logging

from agent.planner import Task

logger = logging.getLogger(__name__)


def dispatch_tasks(tasks: list[Task]) -> list[dict]:
    """Execute each task by dispatching to the appropriate tool."""
    results: list[dict] = []

    for task in tasks:
        if not isinstance(task, dict):
            logger.warning("Skipping non-dict task: %r", task)
            continue

        action = task.get("action")
        if not isinstance(action, str):
            logger.warning("Skipping task with missing/invalid 'action': %r", task)
            results.append({
                "name": "",
                "status": "error",
                "message": "Task missing valid 'action' key",
            })
            continue

        results.append({
            "name": task.get("name", ""),
            "status": "skipped",
            "message": f"Unknown action: {action}",
        })

    return results
